File: src/libs/sdk/dotnet/Ums.Sdk.Contracts/SchemaVersion.cs
namespace Ums.Sdk.Contracts;

/// <summary>
/// Compile-time constants describing the AuthorizationGraph schema this SDK was built against.
/// See ADR-0074 for the versioning policy and SCHEMA_VERSIONING.md for the operational summary.
/// </summary>
public static class SchemaVersion
{
    /// <summary>Current canonical schema version emitted and consumed by SDK 1.0.x.</summary>
    public const string Current = "1.0.0";

    /// <summary>Inclusive lower bound of the schema compatibility range.</summary>
    public const string CompatibilityMinInclusive = "1.0.0";

    /// <summary>Exclusive upper bound of the schema compatibility range.</summary>
    public const string CompatibilityMaxExclusive = "2.0.0";

    /// <summary>
    /// Returns true when <paramref name="schemaVersion"/> falls within the SDK's supported range.
    /// </summary>
    public static bool IsSupported(string? schemaVersion)
    {
        if (string.IsNullOrWhiteSpace(schemaVersion)) return false;
        if (!TryParse(schemaVersion, out var v)) return false;
        if (!TryParse(CompatibilityMinInclusive, out var min)) return false;
        if (!TryParse(CompatibilityMaxExclusive, out var max)) return false;
        return Compare(v, min) >= 0 && Compare(v, max) < 0;
    }

    /// <summary>True when the major component of <paramref name="schemaVersion"/> equals the current major.</summary>
    public static bool IsMajorMatch(string? schemaVersion)
    {
        if (!TryParse(schemaVersion, out var v)) return false;
        if (!TryParse(Current, out var c)) return false;
        return v.Major == c.Major;
    }

    /// <summary>True when <paramref name="schemaVersion"/> is a MINOR ahead of the current schema (server newer).</summary>
    public static bool IsMinorAhead(string? schemaVersion)
    {
        if (!TryParse(schemaVersion, out var v)) return false;
        if (!TryParse(Current, out var c)) return false;
        if (v.Major != c.Major) return false;
        return v.Minor > c.Minor;
    }

    /// <summary>True when <paramref name="schemaVersion"/> is a MINOR behind the current schema (SDK newer).</summary>
    public static bool IsMinorBehind(string? schemaVersion)
    {
        if (!TryParse(schemaVersion, out var v)) return false;
        if (!TryParse(Current, out var c)) return false;
        if (v.Major != c.Major) return false;
        return v.Minor < c.Minor;
    }

    internal static bool TryParse(string? input, out (int Major, int Minor, int Patch) parsed)
    {
        parsed = default;
        if (string.IsNullOrWhiteSpace(input)) return false;
        var parts = input.Split('.');
        if (parts.Length != 3) return false;
        if (!int.TryParse(parts[0], out var major) ||
            !int.TryParse(parts[1], out var minor) ||
            !int.TryParse(parts[2], out var patch)) return false;
        parsed = (major, minor, patch);
        return true;
    }

    internal static int Compare((int Major, int Minor, int Patch) a, (int Major, int Minor, int Patch) b)
    {
        if (a.Major != b.Major) return a.Major.CompareTo(b.Major);
        if (a.Minor != b.Minor) return a.Minor.CompareTo(b.Minor);
        return a.Patch.CompareTo(b.Patch);
    }
}

File: src/libs/sdk/typescript/sdk-contracts/src/schema-version.ts
/**
 * Compile-time constants describing the AuthorizationGraph schema this SDK was built against.
 * See ADR-0074 for the versioning policy.
 */
export const SchemaVersion = {
  Current: '1.0.0',
  CompatibilityMinInclusive: '1.0.0',
  CompatibilityMaxExclusive: '2.0.0'
} as const;

interface Semver {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
}

function tryParse(input: string | null | undefined): Semver | null {
  if (input === null || input === undefined) return null;
  const trimmed = input.trim();
  if (trimmed.length === 0) return null;
  const parts = trimmed.split('.');
  if (parts.length !== 3) return null;
  const major = Number.parseInt(parts[0]!, 10);
  const minor = Number.parseInt(parts[1]!, 10);
  const patch = Number.parseInt(parts[2]!, 10);
  if (!Number.isFinite(major) || !Number.isFinite(minor) || !Number.isFinite(patch)) return null;
  if (String(major) !== parts[0] || String(minor) !== parts[1] || String(patch) !== parts[2]) return null;
  return { major, minor, patch };
}

function compare(a: Semver, b: Semver): number {
  if (a.major !== b.major) return a.major - b.major;
  if (a.minor !== b.minor) return a.minor - b.minor;
  return a.patch - b.patch;
}

/** Returns true when `version` falls within the SDK's supported range. */
export function isSchemaVersionSupported(version: string | null | undefined): boolean {
  const v = tryParse(version);
  if (v === null) return false;
  const min = tryParse(SchemaVersion.CompatibilityMinInclusive)!;
  const max = tryParse(SchemaVersion.CompatibilityMaxExclusive)!;
  return compare(v, min) >= 0 && compare(v, max) < 0;
}

/** True when the MAJOR of `version` matches the SDK's current MAJOR. */
export function isMajorMatch(version: string | null | undefined): boolean {
  const v = tryParse(version);
  if (v === null) return false;
  const current = tryParse(SchemaVersion.Current)!;
  return v.major === current.major;
}

/** True when `version` is a MINOR ahead of the current schema (server newer than SDK). */
export function isMinorAhead(version: string | null | undefined): boolean {
  const v = tryParse(version);
  if (v === null) return false;
  const current = tryParse(SchemaVersion.Current)!;
  if (v.major !== current.major) return false;
  return v.minor > current.minor;
}

/** True when `version` is a MINOR behind the current schema (SDK newer than server). */
export function isMinorBehind(version: string | null | undefined): boolean {
  const v = tryParse(version);
  if (v === null) return false;
  const current = tryParse(SchemaVersion.Current)!;
  if (v.major !== current.major) return false;
  return v.minor < current.minor;
}
